File: js/cr-whitelist-browser.py
#!/usr/bin/env python3
"""
ConnectReseller Auto IP Whitelist via Browser Automation (Playwright)

Since ConnectReseller has no API for IP whitelisting, this script:
1. Detects the server's outbound IP
2. Logs into the ConnectReseller panel
3. Navigates to Tools → Profile → API tab
4. Reads existing whitelisted IPs
5. Adds the server IP to the first empty slot (if not already present)
6. Saves the whitelist

Usage: /opt/plugins-venv/bin/python3 cr-whitelist-browser.py [IP_TO_WHITELIST]
  If no IP provided, auto-detects via ipify.org

Exit codes:
  0 = success (IP whitelisted or already present)
  1 = error (login failed, no empty slots, etc.)

Output: JSON on stdout with { success, ip, message }
"""

import sys
import os
import json
import urllib.request

def detect_ip():
    """Detect outbound IP via ipify"""
    try:
        req = urllib.request.urlopen('https://api.ipify.org/', timeout=5)
        return req.read().decode().strip()
    except Exception:
        try:
            req = urllib.request.urlopen('https://ifconfig.me/ip', timeout=5)
            return req.read().decode().strip()
        except Exception:
            return None

def result(success, ip='', message=''):
    print(json.dumps({'success': success, 'ip': ip, 'message': message}))
    sys.exit(0 if success else 1)

def main():
    email = os.environ.get('CR_PANEL_EMAIL', '')
    password = os.environ.get('CR_PANEL_PASSWORD', '')

    if not email or not password:
        result(False, '', 'CR_PANEL_EMAIL and CR_PANEL_PASSWORD env vars required')

    # Get IP to whitelist
    ip = sys.argv[1] if len(sys.argv) > 1 else detect_ip()
    if not ip:
        result(False, '', 'Could not detect outbound IP')

    from playwright.sync_api import sync_playwright

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={'width': 1920, 'height': 900})
        page = context.new_page()

        try:
            # 1. Login
            page.goto('https://global.connectreseller.com', wait_until='networkidle', timeout=30000)
            page.wait_for_timeout(2000)

            page.fill('input[type="text"]', email)
            page.fill('input[type="password"]', password)
            page.click('button:has-text("Sign in")', force=True)
            page.wait_for_timeout(8000)

            if '/dashboard' not in page.url and '/reseller' not in page.url:
                result(False, ip, f'Login failed - landed on {page.url}')

            # 2. Navigate to Profile → API tab
            page.goto('https://global.connectreseller.com/tools/profile', wait_until='networkidle', timeout=15000)
            page.wait_for_timeout(3000)
            page.click('text=API', force=True)
            page.wait_for_timeout(4000)

            # 3. Read current IP values
            ip_fields = ['ipaddress1', 'ipaddress2', 'ipaddress3', 'ipaddress4', 'ipaddress5']
            current_values = {}
            for name in ip_fields:
                try:
                    val = page.input_value(f'input[formcontrolname="{name}"]')
                    current_values[name] = val.strip() if val else ''
                except Exception:
                    current_values[name] = ''

            existing_ips = [v for v in current_values.values() if v]

            # 4. Check if IP already whitelisted
            if ip in existing_ips:
                result(True, ip, f'IP {ip} is already whitelisted (slots: {existing_ips})')

            # 5. Find first empty slot
            empty_slot = None
            for name in ip_fields:
                if not current_values[name]:
                    empty_slot = name
                    break

            if not empty_slot:
                # Try to find a slot we can replace (not the first one which is primary)
                result(False, ip, f'No empty IP slots. Current: {existing_ips}. Remove one manually.')

            # 6. Fill the empty slot
            page.fill(f'input[formcontrolname="{empty_slot}"]', ip)
            page.wait_for_timeout(1000)

            # 7. Click Save
            page.evaluate('window.scrollTo(0, document.body.scrollHeight)')
            page.wait_for_timeout(500)
            save_btn = page.query_selector('button:has-text("Save whitelisted IP address")')
            if not save_btn:
                result(False, ip, 'Save button not found on page')

            save_btn.click(force=True)
            page.wait_for_timeout(5000)

            # 8. Verify success message
            content = page.content()
            if 'successfully updated' in content.lower() or 'success' in content.lower():
                result(True, ip, f'IP {ip} whitelisted in {empty_slot}')
            else:
                result(True, ip, f'IP {ip} added to {empty_slot} (save clicked, check panel to confirm)')

        except Exception as e:
            result(False, ip, f'Browser automation error: {str(e)}')
        finally:
            browser.close()

if __name__ == '__main__':
    main()

File: js/cr-whitelist-browser.js
/**
 * ConnectReseller Auto IP Whitelist via Browser Automation (Puppeteer)
 *
 * Since ConnectReseller has no API for IP whitelisting, this script:
 * 1. Logs into the ConnectReseller panel
 * 2. Navigates to Tools → Profile → API tab
 * 3. Reads existing whitelisted IPs
 * 4. Adds the server IP to the first empty slot (if not already present)
 * 5. Saves the whitelist
 *
 * Usage: node cr-whitelist-browser.js <IP_TO_WHITELIST>
 * Output: JSON on stdout with { success, ip, message }
 *
 * Configuration:
 * - Set BROWSER_WS_ENDPOINT env var to use external browser service (e.g., Browserless.io)
 * - For local dev without external service, it will attempt local browser launch (requires full puppeteer)
 */

const puppeteer = require('puppeteer-core')
const fs = require('fs')

const IP_FIELDS = ['ipaddress1', 'ipaddress2', 'ipaddress3', 'ipaddress4', 'ipaddress5']

/**
 * Find a working Chromium binary for local development.
 * - On preview pods (aarch64): Playwright-installed Chromium at /pw-browsers/
 * - Fallback to system Chrome/Chromium
 */
function findChromePath() {
  const pwPaths = [
    '/pw-browsers/chromium-1208/chrome-linux/chrome',
    '/pw-browsers/chromium_headless_shell-1208/chrome-linux/headless_shell',
  ]
  for (const p of pwPaths) {
    if (fs.existsSync(p)) return p
  }
  // Common system paths
  const systemPaths = [
    '/usr/bin/chromium-browser',
    '/usr/bin/chromium',
    '/usr/bin/google-chrome',
  ]
  for (const p of systemPaths) {
    if (fs.existsSync(p)) return p
  }
  return undefined
}

function output(success, ip, message) {
  console.log(JSON.stringify({ success, ip, message }))
  process.exit(success ? 0 : 1)
}

async function main() {
  const email = process.env.CR_PANEL_EMAIL || ''
  const password = process.env.CR_PANEL_PASSWORD || ''
  const ip = process.argv[2] || ''

  if (!email || !password) {
    return output(false, ip, 'CR_PANEL_EMAIL and CR_PANEL_PASSWORD env vars required')
  }
  if (!ip) {
    return output(false, '', 'IP address argument required')
  }

  let browser
  try {
    // Use external browser service if BROWSER_WS_ENDPOINT is set (production/Railway)
    const browserWsEndpoint = process.env.BROWSER_WS_ENDPOINT
    
    if (browserWsEndpoint) {
      // Connect to external browser service (Browserless, etc.)
      browser = await puppeteer.connect({
        browserWSEndpoint: browserWsEndpoint,
      })
    } else {
      // Local development fallback - launch local browser
      const chromePath = findChromePath()
      if (!chromePath) {
        return output(false, ip, 'No BROWSER_WS_ENDPOINT set and no local Chrome found. Set BROWSER_WS_ENDPOINT to use external browser service.')
      }
      browser = await puppeteer.launch({
        headless: true,
        executablePath: chromePath,
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'],
      })
    }
    
    const page = await browser.newPage()
    await page.setViewport({ width: 1920, height: 900 })

    // 1. Login
    await page.goto('https://global.connectreseller.com', { waitUntil: 'networkidle2', timeout: 30000 })
    await new Promise(r => setTimeout(r, 2000))

    // Clear fields first, then type credentials
    const emailField = await page.$('input[type="text"]')
    const passField = await page.$('input[type="password"]')
    if (emailField) {
      await emailField.click({ clickCount: 3 })
      await emailField.type(email, { delay: 20 })
    }
    if (passField) {
      await passField.click({ clickCount: 3 })
      await passField.type(password, { delay: 20 })
    }

    // Click Sign in button specifically
    const signInBtn = await page.evaluateHandle(() => {
      const btns = [...document.querySelectorAll('button')]
      return btns.find(b => b.textContent.toLowerCase().includes('sign in'))
    })
    if (signInBtn && await signInBtn.asElement()) {
      await signInBtn.click()
    } else {
      // Fallback: submit the form
      await page.keyboard.press('Enter')
    }
    await new Promise(r => setTimeout(r, 10000))

    const url = page.url()
    if (!url.includes('/dashboard') && !url.includes('/reseller')) {
      return output(false, ip, `Login failed - landed on ${url}`)
    }

    // 2. Navigate to Profile → API tab
    await page.goto('https://global.connectreseller.com/tools/profile', { waitUntil: 'networkidle2', timeout: 15000 })
    await new Promise(r => setTimeout(r, 3000))

    // Click API tab
    const apiTab = await page.evaluateHandle(() => {
      const els = [...document.querySelectorAll('a, button, span, div, li')]
      return els.find(el => el.textContent.trim() === 'API')
    })
    if (apiTab) await apiTab.click()
    await new Promise(r => setTimeout(r, 4000))

    // 3. Read current IP values
    const currentValues = {}
    for (const name of IP_FIELDS) {
      try {
        const val = await page.$eval(
          `input[formcontrolname="${name}"]`,
          el => el.value
        )
        currentValues[name] = (val || '').trim()
      } catch {
        currentValues[name] = ''
      }
    }

    const existingIps = Object.values(currentValues).filter(Boolean)

    // 4. Check if IP already whitelisted
    if (existingIps.includes(ip)) {
      return output(true, ip, `IP ${ip} is already whitelisted (slots: ${existingIps.join(', ')})`)
    }

    // 5. Find first empty slot
    let emptySlot = null
    for (const name of IP_FIELDS) {
      if (!currentValues[name]) {
        emptySlot = name
        break
      }
    }

    if (!emptySlot) {
      return output(false, ip, `No empty IP slots. Current: ${existingIps.join(', ')}. Remove one manually.`)
    }

    // 6. Fill the empty slot
    const field = await page.$(`input[formcontrolname="${emptySlot}"]`)
    await field.click({ clickCount: 3 }) // select all existing text
    await field.type(ip)
    await new Promise(r => setTimeout(r, 1000))

    // 7. Click Save
    await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight))
    await new Promise(r => setTimeout(r, 500))

    const saveBtn = await page.evaluateHandle(() => {
      const btns = [...document.querySelectorAll('button')]
      return btns.find(b => b.textContent.toLowerCase().includes('save whitelisted ip'))
    })
    if (!saveBtn || !(await saveBtn.asElement())) {
      return output(false, ip, 'Save button not found on page')
    }

    await saveBtn.click()
    await new Promise(r => setTimeout(r, 5000))

    // 8. Verify success
    const content = await page.content()
    const lower = content.toLowerCase()
    if (lower.includes('successfully updated') || lower.includes('success')) {
      return output(true, ip, `IP ${ip} whitelisted in ${emptySlot}`)
    } else {
      return output(true, ip, `IP ${ip} added to ${emptySlot} (save clicked, check panel to confirm)`)
    }
  } catch (e) {
    return output(false, ip, `Browser automation error: ${e.message}`)
  } finally {
    if (browser) await browser.close()
  }
}

main()
